packages/user-access-policy/src/rule/storage/indexes/unique/testUniqueIndexBase.ts

```diff
@@ -0,0 +1,145 @@
+// Copyright 2021-2024 Prosopo (UK) Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+import { expect } from "vitest";
+import type Ip from "../../../../ip/ip.js";
+import TestRulesStorageBase from "../../test/testRulesStorageBase.js";
+
+abstract class TestUniqueIndexBase extends TestRulesStorageBase {
+	protected abstract getFirstUserIpObject(): Ip;
+
+	protected abstract getSecondUserIpObject(): Ip;
+
+	protected override getTests(): {
+		name: string;
+		method: () => Promise<void>;
+	}[] {
+		return [
+			{
+				name: "globalRuleAcceptsUniqueIp",
+				method: async () => this.globalRuleAcceptsUniqueIp(),
+			},
+			{
+				name: "globalRuleRejectsNotUniqueIp",
+				method: async () => this.globalRuleRejectsNotUniqueIp(),
+			},
+			{
+				name: "clientRuleAcceptsUniqueIp",
+				method: async () => this.clientRuleAcceptsUniqueIp(),
+			},
+			{
+				name: "clientRuleRejectsNotUniqueIp",
+				method: async () => this.clientRuleRejectsNotUniqueIp(),
+			},
+			{
+				name: "globalAndClientRulesCanHaveSameIp",
+				method: async () => this.globalAndClientRulesCanHaveSameIp(),
+			},
+		];
+	}
+
+	protected async globalRuleAcceptsUniqueIp(): Promise<void> {
```

Comment:
Are these tests? if so, they should be in a tests folder and have a .test extension as per previous comments